server/src/main/java/org/elasticsearch/index/mapper/Mapper.java

```diff
@@ -22,6 +24,57 @@
 
 public abstract class Mapper implements ToXContentFragment, Iterable<Mapper> {
 
+    public static final NodeFeature STORE_SOURCE_MAPPER_PARAM = new NodeFeature("mapper.source.store_param");
+
+    static final String STORE_SOURCE_PARAM = "store_source";
+
+    // Only relevant for synthetic source mode.
+    public enum StoreSourceMode {
+        NONE("none"),      // No source recording
+        ARRAYS("arrays"),  // Store source for arrays of mapped fields
+        ALL("all");        // Store source for both singletons and arrays of mapped fields
+
+        StoreSourceMode(String name) {
+            this.name = name;
+        }
+
+        static StoreSourceMode from(String input) {
+            if (input.equals(NONE.name)) {
+                return NONE;
+            }
+            if (input.equals(ALL.name)) {
+                return ALL;
+            }
+            if (input.equals(ARRAYS.name)) {
+                return ARRAYS;
+            }
+            throw new IllegalArgumentException("Unknown " + STORE_SOURCE_PARAM + " value [" + input + "]");
+        }
+
+        @Override
+        public String toString() {
+            return name;
+        }
+
+        private final String name;
+    }
+
+    // Only relevant for indexes configured with synthetic source mode. Otherwise, it has no effect.
+    // Controls the default behavior for storing the source of leaf fields and objects, in singleton or array form.
+    // Setting to StoreSourceMode.ALL is equivalent to disabling synthetic source, so this is not allowed.
+    public static final Setting<StoreSourceMode> STORE_ARRAY_SOURCE_SETTING = Setting.enumSetting(
+        StoreSourceMode.class,
+        "index.mapping.store_source",
+        StoreSourceMode.NONE,
```

Comment:
Should the default by arrays if `index.mode` is `logsdb`?

Reply:
I think we should change this in the logs template, so that it can be overriden for `logsdb` indexes, if not desired.

Reply:
I see, I didn't know this was the plan to unable. This makes sense to me.